### ISSUE TITLE
Parameters out-of-sync in multi-GPU training, only 1st GPU actually contributes to training

### ISSUE THREAD

**Issue by parlance-zz** [huggingface/accelerate#2907]
Parameters at the beginning of a new training run are identical as verified by the following code:

```
optimizer.step()

...

if accelerator.sync_gradients:
	num_params = num_matches = 0
	for param in module.parameters():
		all_params = accelerator.gather(param.unsqueeze(0))
		num_matches += int(torch.allclose(param.unsqueeze(0), all_params, rtol=1e-3, atol=1e-3))
		num_params += 1
	print(f" pid:{accelerator.local_process_index} num_matches: {num_matches} / {num_params}")

```

On the first step the match count is identical to the total count, but then begins falling quite quickly and reaches 0 matches within ~20 steps or so. 

If the parameters are supposed to be identical on each GPU after gradient synchronization up to float32 precision, why do I see radically different results with 6 GPUs and 1 grad accum step, vs 1 GPU and 6 grad accum steps?

Only after much pain and wasted time did I finally reach the point where I needed to verify if synchronization was actually working, and it turns out it isn't.

There's nothing particularly special about my training code, accelerate is configured and the accelerator and model are prepared properly, in fact it's just a slightly modified version of the generic [example training code](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py) in the diffusers library.

Edit: If I use the same code as above to check the gradients when sync_gradients is set they _do_ in fact match, but the model parameters drift out of sync regardless. If I add code to explicitly sync the model parameters myself via brute force:

```
for param in module.parameters():
    all_params = accelerator.gather(param.unsqueeze(0))
    param.data.copy_(all_params.mean(dim=0))

```
... then I do see the expected sensible results.
